### PR TITLE
feat: Component Expiration Control, Callbacks and Global Components

### PR DESCRIPTION
**Pull request by GabrielTK**
_Please note that this PR does not introduce any breaking changes._
_All changes are also properly documented in JSDoc_

----

This PR adds a few extras into the components system that may be useful in a plenitude of situations:

1. **Custom Expirations & Callbacks:**

This allows developers to set a custom time for commands to expire, as well as specify an expiration callback, therefore allowing the user to be notified when the component is no longer valid. _The new features are introduced as optional parameters, therefore, no breaking changes are introduced._

```typescript
module.exports = class ExpiringCommand extends SlashCommand {
  constructor(creator: SlashCreator) {
    super(creator, {
      name: 'expiring',
      description: 'Creates a button that expires in 10 seconds',
      options: []
    });

  }


  async run(ctx: CommandContext) {
    await ctx.defer();

    let message = await ctx.send("Press the button in 10 seconds!", {components: [{type: ComponentType.ACTION_ROW, components: [{type: ComponentType.BUTTON,
      custom_id: "expiring-btn",
      label: 'Accept Rules',
      style: ButtonStyle.PRIMARY,
    }]}]});
    ctx.registerComponent(
      'expiring-btn', // 1st parameter, custom id - no changes introduced
      (interact)=>{ // 2nd parameter, callback function - no changes introduced
        console.log("Replied!")
        interact.sendFollowUp("This button is yet to expire!");
      },
      10000, //* 3rd Parameter: Optional timeout in milliseconds - NEW
      ()=>{ //* 4th Parameter: Optional timeout callback - NEW
        ctx.send("This button has expired!");
        (message as Message).delete();
      })



  }
}

```

2. **A new system for registering components globally.**
This allows developers to create components that will react consistently whenever the bot is running, no matter how much time has passed, or if it has been rebooted. As an example, this might be useful to interactively add buttons to a server's bulletin board or welcome message.
Global components are registered by calling `registerGlobalComponent` in the `SlashCreator` object, therefore, they can be created:
* In the constructor of the command that emits the components containing them (my preferred way, for organization, as demonstrated below)
* Anywhere else, through the `SlashCreator` object.

```typescript
module.exports = class HelloCommand extends SlashCommand {
  GLOBAL_BUTTON: string; // Stores the Custom ID for the button.
  constructor(creator: SlashCreator) {
    super(creator, {
      name: 'eternity',

      description: 'Says hello to you.',
      options: []
    });
    this.GLOBAL_BUTTON = creator.registerGlobalComponent('hello', (interact)=>{
      interact.sendFollowUp("This button will never expire!");
    })
  }


  async run(ctx: CommandContext) {

    return ctx.send("Press the button to agree to the rules", {components: [{type: ComponentType.ACTION_ROW, components: [{type: ComponentType.BUTTON,
      custom_id: this.GLOBAL_BUTTON,
      label: 'Accept Rules',
      style: ButtonStyle.PRIMARY,
    }]}]});

  }
}

```

### CODE REVIEW

File: src/creator.ts [hunk 1]
```diff
@@ -238,6 +238,23 @@ export class SlashCreator extends (EventEmitter as any as new () => TypedEventEm
...
+  /**
+   * Registers a global component callback. Note that this will have no expiration, and should be invoked by the returned name.
+   * @param custom_id The custom ID of the component to register
+   * @param callback The callback to use on interaction
+   */
+  registerGlobalComponent(custom_id: string, callback: ComponentRegisterCallback) {
+    if (this._componentCallbacks.has(custom_id))
+      throw new Error(`A global component with the ID "${custom_id}" is already registered.`);
+    const newName = `global-${custom_id}`;
+    this._componentCallbacks.set(newName, {
```
Comment (sudojunior) [inline src/creator.ts:250]: If you're declaring a new identifier - shouldn't that be put above everything else to be called on by the precondition?
- L247 - checks with `custom_id`
  - L248 - throws if true with `custom_id` in the reason
- L249 - declare `newName` using `custom_id`
- L250 - register the callback with the internal map **using `newName`**

File: src/creator.ts [hunk 2]
```diff
@@ -516,7 +533,10 @@ export class SlashCreator extends (EventEmitter as any as new () => TypedEventEm
   cleanRegisteredComponents() {
     if (this._componentCallbacks.size)
       for (const [key, callback] of this._componentCallbacks) {
-        if (callback.expires < Date.now()) this._componentCallbacks.delete(key);
+        if (callback.expires != null && callback.expires < Date.now()) {
+          if (callback.onExpired != null) callback.onExpired();
```
Comment (sudojunior) [inline src/creator.ts:537]: This could use optional chaining... but I'm not exactly sure on the implications.
```suggestion
          callback.onExpired?.();
```

**Review by sudojunior — CHANGES_REQUESTED** [review]
I've done my best to look over what I know, as well as looking at the logical context of a component interaction.

File: src/creator.ts (hunk 2 above)
Comment (GabrielTK) [reply src/creator.ts:537]: Unfortunately, changing it to use optional chaining would break the unit tests.
For more information, please refer to [these logs.](https://github.com/Snazzah/slash-create/runs/3763097018)

File: src/creator.ts (hunk 1 above)
Comment (GabrielTK) [reply src/creator.ts:250]: I'm sorry, but I am unable to comprehend your suggestion or what it would change on the code mentioned. Would you mind demonstrating what you mean?

File: src/creator.ts (hunk 2 above)
Comment (Snazzah) [reply src/creator.ts:537]: Having an `onExpired` function would imply that there is a timer for components, when it only cleans during each component interaction. Not too sure how this would help when it's not on time.

File: src/creator.ts [hunk 3]
```diff
@@ -238,6 +238,23 @@ export class SlashCreator extends (EventEmitter as any as new () => TypedEventEm
...
 
+  /**
+   * Registers a global component callback. Note that this will have no expiration, and should be invoked by the returned name.
+   * @param custom_id The custom ID of the component to register
+   * @param callback The callback to use on interaction
+   */
+  registerGlobalComponent(custom_id: string, callback: ComponentRegisterCallback) {
+    if (this._componentCallbacks.has(custom_id))
+      throw new Error(`A global component with the ID "${custom_id}" is already registered.`);
+    const newName = `global-${custom_id}`;
```
Comment (Snazzah) [inline src/creator.ts:249]:
```suggestion
  registerGlobalComponent(custom_id: string, callback: ComponentRegisterCallback) {
    const newName = `global-${custom_id}`;
    if (this._componentCallbacks.has(newName))
      throw new Error(`A global component with the ID "${newName}" is already registered.`);
```

**Review by Snazzah — CHANGES_REQUESTED** [review]
Not sure about adding `onExpired`, read comments.

File: src/creator.ts (hunk 3 above)
Comment (GabrielTK) [reply src/creator.ts:249]: I'm sorry, shouldn't have submitted that the way it is. Guess I should drink more coffee before opening a PR :P

File: src/creator.ts (hunk 2 above)
Comment (GabrielTK) [reply src/creator.ts:537]: I can provide an example of when the `onExpired` callback would be helpful:
I currently maintain a bot that allows users to query their Genshin Impact in-game user profile. It generates a dropdown to allow users to select their characters to view more details. However, after 15min, the dropdown still stays in the chat but stops reacting to messages since its callback is not registered anymore. Having an `onExpired` callback would allow the bot to send a "final" message or edit the original bot response to replace the components with a "Please rerun the command to interact with it" message.

TLDR: It would allow fewer "hanging" and unreactive components on the chat after the callback's timer has elapsed.

**Review by Snazzah — CHANGES_REQUESTED** [review]
One more thing, should update documentation with examples on this new feature in `docs/examples/components.md`, probably just for globally setting a component.

**Comment by GabrielTK** [comment]
I'll work on that right away.

**Comment by Snazzah** [comment]
Yes, you should also make a disclaimer in the docs stating that the expiration is only checked over when a component interaction occurs.

**Comment by GabrielTK** [comment]
@Snazzah Docs added :)
